### PR TITLE
fix: prefetch state checks

### PR DESCRIPTION
**Pull request by DanielSchiavini**
### What I did
Fix the prefetch fork state

### How I did it
- Fixed the `is_forked` property of PyEVM class
- Avoid locked changes, use commit instead
- Added tests

### How to verify it
Test are included

### Description for the changelog
- fix: prefetch state checks

### Cute Animal Picture
image

### CODE REVIEW

File: boa/vm/fork.py [hunk 1]
```diff
@@ -256,7 +256,7 @@ def try_prefetch_state(self, msg: Message):
                 key = int_to_big_endian(slot)
                 if not self._helper_have_storage(address, slot):
                     account_store._journal_storage[key] = rlp.encode(value)  # type: ignore
-        self.lock_changes()
```
Comment (DanielSchiavini) [inline boa/vm/fork.py:259]: @charles-cooper is this correct? `lock_changes` was invalidating the snapshot taken in the boa pytest plugin, crashing the test runner

Comment (charles-cooper) [reply boa/vm/fork.py:259]: the idea behind using `lock_changes()` was that the items from the RPC should be the "pre-fork state".

we could use a different technique, but i don't think using `commit(snapshot)` is ideal because the fetched RPC state would get rolled back after `anchor()`

File: tests/integration/fork/test_from_etherscan.py [hunk 2]
```diff
@@ -33,3 +36,36 @@ def test_proxy_contract(proxy_contract):
...
+    assert (
+        db.get_code(proxy_contract.address.canonical_address)
+        == proxy_contract._bytecode
+    )
+
+
+@pytest.mark.parametrize("prefetch", [True, False])
+def test_prefetch_state_called_on_message(proxy_contract, prefetch):
+    boa.env.evm._fork_try_prefetch_state = prefetch
+    with patch("boa.vm.fork.AccountDBFork.try_prefetch_state") as mock:
```
Comment (charles-cooper) [inline tests/integration/fork/test_from_etherscan.py:61]: neat

File: boa/vm/fork.py (hunk 1 above)
Comment (DanielSchiavini) [reply boa/vm/fork.py:259]: as discussed elsewhere, the code is now adapted to cache the RPC call

File: boa/vm/fork.py [hunk 3]
```diff
@@ -236,14 +236,18 @@ def try_prefetch_state(self, msg: Message):
                 return
 
             # set account if we don't already have it
-            if self._get_account_helper(address) is None:
-                balance = to_int(v.get("balance", "0x"))
-                code = to_bytes(v.get("code", "0x"))
-                nonce = v.get("nonce", 0)  # already an int
+            account_helper = self._get_account_helper(address)
+            if account_helper is None:
```
Comment (charles-cooper) [inline boa/vm/fork.py:240]: i think we should check `dontfetch` instead, it's designed so we don't have to peer into pyevm internals

File: boa/vm/fork.py [hunk 4]
```diff
@@ -214,20 +219,15 @@ def try_prefetch_state(self, msg: Message):
...
+            trace_args = [args, self._block_id, {"tracer": "prestateTracer"}]
+            trace = self._rpc.fetch("debug_traceCall", trace_args)
         except (RPCError, HTTPError):
             return
 
         snapshot = self.record()
 
-        # everything is returned in hex
-        for address, v in res.items():
+        for address, account_trace in trace.items():
```
Comment (charles-cooper) [inline boa/vm/fork.py:230]: account_dict?

File: boa/vm/fork.py [hunk 5]
```diff
@@ -237,35 +237,36 @@ def try_prefetch_state(self, msg: Message):
...
-                balance = to_int(v.get("balance", "0x"))
-                code = to_bytes(v.get("code", "0x"))
-                nonce = v.get("nonce", 0)  # already an int
+                balance = to_int(account_trace.get("balance", "0x"))
+                code = to_bytes(account_trace.get("code", "0x"))
+                nonce = account_trace.get("nonce", 0)  # already an int
                 self._set_account(address, Account(nonce=nonce, balance=balance))
                 self.set_code(address, code)
 
-            storage = v.get("storage", dict())
```
Comment (charles-cooper) [inline boa/vm/fork.py:246]: i think we can keep this line

File: boa/vm/fork.py [hunk 6]
```diff
@@ -237,35 +237,36 @@ def try_prefetch_state(self, msg: Message):
...
                 # see AccountStorageDB.get()
                 # note we explicitly write 0s, so that they appear
                 # in the journal later when called by get_storage
-                key = int_to_big_endian(slot)
                 if not self._helper_have_storage(address, slot):
-                    account_store._journal_storage[key] = rlp.encode(value)  # type: ignore
-        self.lock_changes()
+                    self.set_storage(address, slot, value)
+
+        # the prefetch is lost on later reverts, however the RPC calls are cached
```
Comment (charles-cooper) [inline boa/vm/fork.py:256]: nice comment

File: boa/vm/fork.py [hunk 7]
```diff
@@ -310,7 +311,8 @@ def set_storage(self, address, slot, value):
         self._dontfetch[key] = _HAS_KEY
 
     def account_exists(self, address):
-        if super().account_exists(address):
-            return True
-
-        return self.get_balance(address) > 0 or self.get_nonce(address) > 0
+        return (
```
Comment (charles-cooper) [inline boa/vm/fork.py:314]: is this change necessary?

Comment (DanielSchiavini) [reply boa/vm/fork.py:314]: not really :grin:

File: boa/vm/fork.py [hunk 8]
```diff
@@ -277,31 +278,31 @@ def commit(self, checkpoint):
...
-        key = self._get_storage_tracker_key(address, slot)
-        return self._dontfetch.get(key) == _HAS_KEY
+        db = super()._get_address_store(address)._locked_changes
+        key = int_to_big_endian(slot)
+        return db.get(key, _EMPTY) != _EMPTY
 
     def get_storage(self, address, slot, from_journal=True):
         # call super for address warming semantics
-        val = super().get_storage(address, slot, from_journal)
+        value = super().get_storage(address, slot, from_journal)
```
[the remark this answers is not in the document]
Comment (DanielSchiavini) [reply boa/vm/fork.py:296]: not really :grinning:

File: boa/vm/fork.py [hunk 9]
```diff
@@ -277,31 +278,31 @@ def commit(self, checkpoint):
 
     def record(self):
         checkpoint = super().record()
-        self._dontfetch.record(checkpoint)
-        return checkpoint
+        return self._dontfetch.record(checkpoint)
```
Comment (charles-cooper) [inline boa/vm/fork.py:281]: is this change necessary? i wouldn't trust that dontfetch.record() returns the same thing as super.record()

Comment (DanielSchiavini) [reply boa/vm/fork.py:281]: `record` must return the checkpoint, and since you've given a custom one it must respect it.
https://github.com/ethereum/py-evm/blob/d8df507e42c885ddeca2b64bbb8f10705f075d3e/eth/db/journal.py#L143

But yeah it's not necessary, I can revert it

Comment (charles-cooper) [reply boa/vm/fork.py]: i mean i'm aware of what the implementation *currently* does, but if it ever changes, we want to match super, not dontfetch.

File: boa/vm/fork.py [hunk 10]
```diff
@@ -283,25 +284,26 @@ def record(self):
     # helper to determine if something is in the storage db
     # or we need to get from RPC
     def _helper_have_storage(self, address, slot, from_journal=True):
-        if not from_journal:
-            db = super()._get_address_store(address)._locked_changes
-            key = int_to_big_endian(slot)
-            return db.get(key, _EMPTY) != _EMPTY
+        if from_journal:
```
[the remark this answers is not in the document]
Comment (DanielSchiavini) [reply boa/vm/fork.py:287]: Double negatives are harder to understand 😁 but I can revert if you don't want it

[the remark this answers is not in the document]
Comment (charles-cooper) [reply boa/vm/fork.py:287]: please revert it, it was clearer to begin with!

**Review by charles-cooper — APPROVED** [review]
thanks!
